README.md
# mopidy-nad

[![Latest PyPI version](https://img.shields.io/pypi/v/mopidy-nad)](https://pypi.org/p/mopidy-nad)
[![CI build status](https://img.shields.io/github/actions/workflow/status/mopidy/mopidy-nad/ci.yml)](https://github.com/mopidy/mopidy-nad/actions/workflows/ci.yml)
[![Test coverage](https://img.shields.io/codecov/c/gh/mopidy/mopidy-nad)](https://codecov.io/gh/mopidy/mopidy-nad)

[Mopidy](https://mopidy.com) extension for controlling volume on a NAD amplifier.
Developed and tested with a NAD C355BEE.


## Maintainer wanted

Mopidy-NAD is currently kept on life support by the Mopidy core developers.
It is in need of a more dedicated maintainer.

If you want to be the maintainer of Mopidy-NAD, please:

1. Make 2-3 good pull requests improving any part of the project.

2. Read and get familiar with all of the project's open issues.

3. Send a pull request removing this section and adding yourself as the
   "Current maintainer" in the "Credits" section below. In the pull request
   description, please refer to the previous pull requests and state that
   you've familiarized yourself with the open issues.

As a maintainer, you'll be given push access to the repo and the authority to
make releases to PyPI when you see fit.


## Installation

Install by running:

```sh
python3 -m pip install mopidy-nad
```

See https://mopidy.com/ext/nad/ for alternative installation methods.


## Configuration

The Mopidy-NAD extension is enabled by default. To disable it, add the
following to `mopidy.conf`:

```ini
[nad]
enabled = false
```

The NAD amplifier must be connected to the machine running Mopidy using a
serial cable.

To use the NAD amplifier to control volume, set the `audio/mixer` config
value in `mopidy.conf` to `nad`. You probably also needs to add some
properties to the `nad` config section.

Supported properties includes:

- `port`: The serial device to use, defaults to `/dev/ttyUSB0`. This must
  be set correctly for the mixer to work.

- `source`: The source that should be selected on the amplifier, like
  `aux`, `disc`, `tape`, `tuner`, etc. Leave unset if you don't want
  the mixer to change it for you.

- `speakers-a`: Set to `on` or `off` (or `true` or `false`) if you
  want the mixer to make sure that speaker set A is turned on or off. Leave
  unset if you don't want the mixer to change it for you.

- `speakers-b`: See `speakers-a`.

Configuration example with minimum configuration, if the amplifier is available
at `/dev/ttyUSB0`::

```ini
[audio]
mixer = nad
```

Configuration example with minimum configuration, if the amplifier is available
elsewhere:

```ini
[audio]
mixer = nad

[nad]
port = /dev/ttyUSB3
```

Configuration example with full configuration::

```ini
[audio]
mixer = nad

[nad]
port = /dev/ttyUSB0
source = aux
speakers-a = true
speakers-b = false
```


## Project resources

- [Source code](https://github.com/mopidy/mopidy-nad)
- [Issues](https://github.com/mopidy/mopidy-nad/issues)
- [Releases](https://github.com/mopidy/mopidy-nad/releases)


## Development

### Set up development environment

Clone the repo using, e.g. using [gh](https://cli.github.com/):

```sh
gh repo clone mopidy/mopidy-nad
```

Enter the directory, and install dependencies using [uv](https://docs.astral.sh/uv/):

```sh
cd mopidy-nad/
uv sync
```

### Running tests

To run all tests and linters in isolated environments, use
[tox](https://tox.wiki/):

```sh
tox
```

To only run tests, use [pytest](https://pytest.org/):

```sh
pytest
```

To format the code, use [ruff](https://docs.astral.sh/ruff/):

```sh
ruff format .
```

To check for lints with ruff, run:

```sh
ruff check .
```

To check for type errors, use [pyright](https://microsoft.github.io/pyright/):

```sh
pyright .
```

### Making a release

To make a release to PyPI, go to the project's [GitHub releases
page](https://github.com/mopidy/mopidy-nad/releases)
and click the "Draft a new release" button.

In the "choose a tag" dropdown, select the tag you want to release or create a
new tag, e.g. `v0.1.0`. Add a title, e.g. `v0.1.0`, and a description of the changes.

Decide if the release is a pre-release (alpha, beta, or release candidate) or
should be marked as the latest release, and click "Publish release".

Once the release is created, the `release.yml` GitHub Action will automatically
build and publish the release to
[PyPI](https://pypi.org/project/mopidy-nad/).


## Credits

- Original author: [Stein Magnus Jodal](https://github.com/jodal)
- Current maintainer: None. Maintainer wanted, see section above.
- [Contributors](https://github.com/mopidy/mopidy-nad/graphs/contributors)


CHANGELOG.rst
*********
Changelog
*********


v3.0.0 (2019-12-22)
===================

- Depend on final release of Mopidy 3.0.0.


v3.0.0rc1 (2019-11-18)
======================

- Require Mopidy >= 3.0.0a5. (PR: #3)

- Require Python >= 3.7. (PR: #3)

- Update project setup. (PR: #3)


v2.0.0 (2014-07-21)
===================

- Require Mopidy >= 0.19 and the new Mopidy mixer API.

- The configuration format has changed, due to the move from GStreamer 0.10's
  mixer API to Mopidy's new mixer API.


v1.1.0 (2014-01-20)
===================

- Require Mopidy >= 0.18.


v1.0.0 (2013-10-08)
===================

- Update to match the ``cookiecutter-mopidy-ext`` project template.


v0.3 (2013-04-16)
=================

- Include ``README.rst`` and ``LICENSE`` in PyPI package.


v0.2 (2013-04-16)
=================

- Add missing ``create_track()`` helper function.


v0.1 (2013-04-16)
=================

- Extracted extension from Mopidy core.


tests/test_mixer.py
import mopidy.mixer
from mopidy_nad import mixer


def test_is_a_mopidy_mixer():
    assert issubclass(mixer.NadMixer, mopidy.mixer.Mixer)


# TODO Add more tests


tests/test_extension.py
from mopidy_nad import Extension


def test_get_default_config():
    ext = Extension()

    config = ext.get_default_config()

    assert "[nad]" in config
    assert "enabled = true" in config
    assert "port = /dev/ttyUSB0" in config
    assert "source =" in config
    assert "speakers-a =" in config
    assert "speakers-b =" in config


def test_get_config_schema():
    ext = Extension()

    schema = ext.get_config_schema()

    assert "enabled" in schema
    assert "port" in schema
    assert "source" in schema
    assert "speakers-a" in schema
    assert "speakers-b" in schema


mopidy_nad/talker.py
import logging

import pykka
import serial

logger = logging.getLogger(__name__)


class NadTalker(pykka.ThreadingActor):
    """
    Independent thread which does the communication with the NAD amplifier.

    Since the communication is done in an independent thread, Mopidy won't
    block other requests while doing rather time consuming work like
    calibrating the NAD amplifier's volume.
    """

    # Serial link config
    BAUDRATE = 115200
    BYTESIZE = 8
    PARITY = 'N'
    STOPBITS = 1

    # Timeout in seconds used for read/write operations.
    # If you set the timeout too low, the reads will never get complete
    # confirmations and calibration will decrease volume forever. If you set
    # the timeout too high, stuff takes more time. 0.2s seems like a good value
    # for NAD C 355BEE.
    TIMEOUT = 0.2

    # Number of volume levels the amplifier supports. 40 for NAD C 355BEE.
    VOLUME_LEVELS = 40

    def __init__(self, port, source, speakers_a, speakers_b):
        super(NadTalker, self).__init__()

        self.port = port
        self.source = source
        self.speakers_a = speakers_a
        self.speakers_b = speakers_b

        # Volume in range 0..VOLUME_LEVELS. :class:`None` before calibration.
        self._nad_volume = None

        self._device = None

    def on_start(self):
        self._open_connection()
        self._set_device_to_known_state()

    def _open_connection(self):
        logger.info('NAD amplifier: Connecting through "%s"', self.port)
        self._device = serial.Serial(
            port=self.port,
            baudrate=self.BAUDRATE,
            bytesize=self.BYTESIZE,
            parity=self.PARITY,
            stopbits=self.STOPBITS,
            timeout=self.TIMEOUT)
        self._get_device_model()

    def _set_device_to_known_state(self):
        self._power_device_on()
        self._select_speakers()
        self._select_input_source()
        self.mute(False)
        self.calibrate_volume()

    def _get_device_model(self):
        model = self._ask_device('Main.Model')
        logger.info('NAD amplifier: Connected to model "%s"', model)
        return model

    def _power_device_on(self):
        self._check_and_set('Main.Power', 'On')

    def _select_speakers(self):
        if self.speakers_a is not None:
            self._check_and_set('Main.SpeakerA', self.speakers_a.title())
        if self.speakers_b is not None:
            self._check_and_set('Main.SpeakerB', self.speakers_b.title())

    def _select_input_source(self):
        if self.source is not None:
            self._check_and_set('Main.Source', self.source.title())

    def mute(self, mute):
        if mute:
            self._check_and_set('Main.Mute', 'On')
        else:
            self._check_and_set('Main.Mute', 'Off')

    def calibrate_volume(self, current_nad_volume=None):
        # The NAD C 355BEE amplifier has 40 different volume levels. We have no
        # way of asking on which level we are. Thus, we must calibrate the
        # mixer by decreasing the volume 39 times.
        if current_nad_volume is None:
            current_nad_volume = self.VOLUME_LEVELS
        if current_nad_volume == self.VOLUME_LEVELS:
            logger.info('NAD amplifier: Calibrating by setting volume to 0')
        self._nad_volume = current_nad_volume
        if self._decrease_volume():
            current_nad_volume -= 1
        if current_nad_volume == 0:
            logger.info('NAD amplifier: Done calibrating')
        else:
            self.actor_ref.proxy().calibrate_volume(current_nad_volume)

    def set_volume(self, volume):
        # Increase or decrease the amplifier volume until it matches the given
        # target volume.
        logger.debug('Setting volume to %d' % volume)
        target_nad_volume = int(round(volume * self.VOLUME_LEVELS / 100.0))
        if self._nad_volume is None:
            return  # Calibration needed
        while target_nad_volume > self._nad_volume:
            if self._increase_volume():
                self._nad_volume += 1
        while target_nad_volume < self._nad_volume:
            if self._decrease_volume():
                self._nad_volume -= 1

    def _increase_volume(self):
        # Increase volume. Returns :class:`True` if confirmed by device.
        self._write('Main.Volume+')
        return self._readline() == 'Main.Volume+'

    def _decrease_volume(self):
        # Decrease volume. Returns :class:`True` if confirmed by device.
        self._write('Main.Volume-')
        return self._readline() == 'Main.Volume-'

    def _check_and_set(self, key, value):
        for attempt in range(1, 4):
            if self._ask_device(key) == value:
                return
            logger.info(
                'NAD amplifier: Setting "%s" to "%s" (attempt %d/3)',
                key, value, attempt)
            self._command_device(key, value)
        if self._ask_device(key) != value:
            logger.info(
                'NAD amplifier: Gave up on setting "%s" to "%s"',
                key, value)

    def _ask_device(self, key):
        self._write('%s?' % key)
        return self._readline().replace('%s=' % key, '')

    def _command_device(self, key, value):
        if type(value) == unicode:
            value = value.encode('utf-8')
        self._write('%s=%s' % (key, value))
        self._readline()

    def _write(self, data):
        # Write data to device. Prepends and appends a newline to the data, as
        # recommended by the NAD documentation.
        if not self._device.isOpen():
            self._device.open()
        self._device.write('\n%s\n' % data)
        logger.debug('Write: %s', data)

    def _readline(self):
        # Read line from device. The result is stripped for leading and
        # trailing whitespace.
        if not self._device.isOpen():
            self._device.open()
        result = self._device.readline().strip()
        if result:
            logger.debug('Read: %s', result)
        return result


mopidy_nad/mixer.py
"""Mixer that controls volume using a NAD amplifier."""

import logging

import pykka

import serial
from mopidy import mixer

logger = logging.getLogger(__name__)


class NadMixer(pykka.ThreadingActor, mixer.Mixer):

    name = "nad"

    def __init__(self, config):
        super().__init__(config)

        self.port = config["nad"]["port"]
        self.source = config["nad"]["source"] or None
        self.speakers_a = config["nad"]["speakers-a"]
        self.speakers_b = config["nad"]["speakers-b"]

        self._volume_cache = 0
        self._mute_cache = False

        self._device = None

        # Volume in range 0..VOLUME_LEVELS, None before calibration.
        self._nad_volume = None

    def get_volume(self):
        return self._volume_cache

    def set_volume(self, volume):
        success = self._set_volume(volume)
        if success:
            self._volume_cache = volume
            self.trigger_volume_changed(volume)
        return success

    def get_mute(self):
        return self._mute_cache

    def set_mute(self, mute):
        success = self._mute(mute)
        if success:
            self._mute_cache = mute
            self.trigger_mute_changed(mute)
        return success

    # Serial link config
    BAUDRATE = 115200
    BYTESIZE = 8
    PARITY = "N"
    STOPBITS = 1

    # Timeout in seconds used for read/write operations.
    # If you set the timeout too low, the reads will never get complete
    # confirmations and calibration will decrease volume forever. If you set
    # the timeout too high, stuff takes more time. 0.2s seems like a good value
    # for NAD C 355BEE.
    TIMEOUT = 0.2

    # Number of volume levels the amplifier supports. 40 for NAD C 355BEE.
    VOLUME_LEVELS = 40

    def on_start(self):
        self._open_connection()
        self._set_device_to_known_state()

    def _open_connection(self):
        logger.info(f"NAD mixer: Connecting through {self.port!r}")
        self._device = serial.Serial(
            port=self.port,
            baudrate=self.BAUDRATE,
            bytesize=self.BYTESIZE,
            parity=self.PARITY,
            stopbits=self.STOPBITS,
            timeout=self.TIMEOUT,
        )
        self._get_device_model()

    def _set_device_to_known_state(self):
        self._power_device_on()
        self._select_speakers()
        self._select_input_source()
        self._mute(False)
        self.calibrate_volume()

    def _get_device_model(self):
        model = self._ask_device("Main.Model")
        logger.info(f"NAD mixer: Connected to model {model!r}")
        return model

    def _power_device_on(self):
        self._check_and_set("Main.Power", "On")

    def _select_speakers(self):
        if self.speakers_a is not None:
            self._check_and_set(
                "Main.SpeakerA", "On" if self.speakers_a else "Off"
            )
        if self.speakers_b is not None:
            self._check_and_set(
                "Main.SpeakerB", "On" if self.speakers_b else "Off"
            )

    def _select_input_source(self):
        if self.source is not None:
            self._check_and_set("Main.Source", self.source.title())

    def _mute(self, mute):
        if mute:
            return self._check_and_set("Main.Mute", "On")
        else:
            return self._check_and_set("Main.Mute", "Off")

    def calibrate_volume(self, current_nad_volume=None):
        # The NAD C 355BEE amplifier has 40 different volume levels. We have no
        # way of asking on which level we are. Thus, we must calibrate the
        # mixer by decreasing the volume 39 times.
        if current_nad_volume is None:
            current_nad_volume = self.VOLUME_LEVELS
        if current_nad_volume == self.VOLUME_LEVELS:
            logger.info("NAD mixer: Calibrating by setting volume to 0")
        self._nad_volume = current_nad_volume
        if self._decrease_volume():
            current_nad_volume -= 1
        if current_nad_volume == 0:
            logger.info("NAD mixer: Done calibrating")
        else:
            self.actor_ref.proxy().calibrate_volume(current_nad_volume)

    def _set_volume(self, volume):
        # Increase or decrease the amplifier volume until it matches the given
        # target volume.
        logger.debug(f"Setting volume to {volume}")
        target_nad_volume = int(round(volume * self.VOLUME_LEVELS / 100.0))
        if self._nad_volume is None:
            return False  # Calibration needed
        while target_nad_volume > self._nad_volume:
            if self._increase_volume():
                self._nad_volume += 1
        while target_nad_volume < self._nad_volume:
            if self._decrease_volume():
                self._nad_volume -= 1
        return True

    def _increase_volume(self):
        # Increase volume. Returns :class:`True` if confirmed by device.
        self._write("Main.Volume+")
        return self._readline() == "Main.Volume+"

    def _decrease_volume(self):
        # Decrease volume. Returns :class:`True` if confirmed by device.
        self._write("Main.Volume-")
        return self._readline() == "Main.Volume-"

    def _check_and_set(self, key, value):
        for attempt in range(1, 4):
            if self._ask_device(key) == value:
                return True
            logger.info(
                f"NAD mixer: Setting {key!r} to {value!r} (attempt {attempt}/3)"
            )
            self._command_device(key, value)
        if self._ask_device(key) == value:
            return True
        else:
            logger.warning(
                f"NAD mixer: Gave up on setting {key!r} to {value!r}"
            )
            return False

    def _ask_device(self, key):
        self._write("%s?" % key)
        return self._readline().replace(f"{key}=", "")

    def _command_device(self, key, value):

        self._write(f"{key}={value}")
        self._readline()

    def _write(self, data: str):
        # Write data to device. Prepends and appends a newline to the data, as
        # recommended by the NAD documentation.
        if not self._device.isOpen():
            self._device.open()
        data_bytes = f"\n{data}\n".encode()
        self._device.write(data_bytes)
        logger.debug(f"Write: {data_bytes!r}")

    def _readline(self) -> str:
        # Read line from device. The result is stripped for leading and
        # trailing whitespace.
        if not self._device.isOpen():
            self._device.open()
        result_bytes = self._device.readline().strip()
        if result_bytes:
            logger.debug(f"Read: {result_bytes!r}")
        result = result_bytes.decode()
        return result
